In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import re


import warnings
warnings.filterwarnings('ignore')

### Using header=1 to remove the 1st row index named as "Bejaia Region Dataset"

In [2]:
df1=pd.read_csv("D:\INeuron\Full Stack data science\Coding\ML\Job_Project\Algerian_forest_fires_dataset.csv",header=1)
df1.head()

day month  year Temperature#  RH!  Ws@ Rain   FFMC  DMC    DC  ISI  BUI  \
0  01    06  2012           29   57   18     0  65.7  3.4   7.6  1.3  3.4   
1  02    06  2012           29   61   13   1.3  64.4  4.1   7.6    1  3.9   
2  03    06  2012           26   82   22  13.1  47.1  2.5   7.1  0.3  2.7   
3  04    06  2012           25   89   13   2.5  28.6  1.3   6.9    0  1.7   
4  05    06  2012           27   77   16     0  64.8    3  14.2  1.2  3.9   

   FWI    Classes    
0  0.5  not fire     
1  0.4  not fire     
2  0.1  not fire     
3    0  not fire     
4  0.5  not fire

### DATASET INFORMATION

The dataset includes 244 instances that regroup a data of two regions of Algeria:
    
      *  Bejaia region located in the northeast of Algeria
      *  Sidi Bel-abbes region located in the northwest of Algeria

122 instances for each region. 

The period from June 2012 to September 2012. 
The dataset includes 11 attribues and 1 output attribue (class)
The 244 instances have been classified into as fired (138 classes) and as not fired (106 classes) classes.

### VARIABLE INFORMATION

1. Date : (DD/MM/YYYY) Day, month ('june' to 'september'), year (2012) 

#### Weather data observations 
2. Temp : temperature noon (temperature max)  in Celsius degrees: 22 to 42
3. RH : Relative Humidity in %: 21 to 90 
4. Ws :Wind speed in km/h: 6 to 29 
5. Rain: total day in mm: 0 to 16.8

#### FWI Components  
6. Fine Fuel Moisture Code (FFMC) index from the FWI system: 28.6 to 92.5 
7. Duff Moisture Code (DMC) index from the FWI system: 1.1 to 65.9 
8. Drought Code (DC) index from the FWI system:  7 to 220.4
9. Initial Spread Index (ISI) index from the FWI system: 0 to 18.5 
10. Buildup Index (BUI) index from the FWI system: 1.1 to 68
11. Fire Weather Index (FWI) Index: 0 to 31.1
12. Classes: two classes, namely as fired and as not fired.

# EDA

In [3]:
df1.head()

day month  year Temperature#  RH!  Ws@ Rain   FFMC  DMC    DC  ISI  BUI  \
0  01    06  2012           29   57   18     0  65.7  3.4   7.6  1.3  3.4   
1  02    06  2012           29   61   13   1.3  64.4  4.1   7.6    1  3.9   
2  03    06  2012           26   82   22  13.1  47.1  2.5   7.1  0.3  2.7   
3  04    06  2012           25   89   13   2.5  28.6  1.3   6.9    0  1.7   
4  05    06  2012           27   77   16     0  64.8    3  14.2  1.2  3.9   

   FWI    Classes    
0  0.5  not fire     
1  0.4  not fire     
2  0.1  not fire     
3    0  not fire     
4  0.5  not fire

In [4]:
df=df1.copy()

In [5]:
df.columns

Index(['day', 'month', 'year', 'Temperature#', ' RH!', ' Ws@', 'Rain ', 'FFMC',
       'DMC', 'DC', 'ISI', 'BUI', 'FWI', 'Classes  '],
      dtype='object')

In [6]:
#Remove the special characters & spaces in the column names
df.columns=[re.sub("['`','~','!','@','#','$','%','^','&','*','(',')',' ']","",feature) for feature in df.columns]
df.columns

Index(['day', 'month', 'year', 'Temperature', 'RH', 'Ws', 'Rain', 'FFMC',
       'DMC', 'DC', 'ISI', 'BUI', 'FWI', 'Classes'],
      dtype='object')

In [7]:
#Remove the special characters & spaces in each column values
for feature in df.columns:
    df[feature]=df[feature].str.strip()

In [8]:
#Check if there is any null values
df.isnull().sum()

day            0
month          1
year           1
Temperature    1
RH             1
Ws             1
Rain           1
FFMC           1
DMC            1
DC             1
ISI            1
BUI            1
FWI            1
Classes        2
dtype: int64

In [9]:
#Create a null_mask to fetch the records having null values
null_mask=df.isnull().any(axis=1)
df[null_mask]

day month  year Temperature   RH   Ws Rain  \
122  Sidi-Bel Abbes Region Dataset   NaN   NaN         NaN  NaN  NaN  NaN   
167                             14    07  2012          37   37   18  0.2   

     FFMC   DMC      DC   ISI   BUI   FWI Classes  
122   NaN   NaN     NaN   NaN   NaN   NaN     NaN  
167  88.9  12.9  14.6 9  12.5  10.4  fire     NaN

In [10]:
df.shape

(246, 14)

In [11]:
# Drop the Null value records
df.drop(index=[122],axis=0,inplace=True)
df.shape

(245, 14)

In [12]:
df[null_mask]

day month  year Temperature  RH  Ws Rain  FFMC   DMC      DC   ISI   BUI  \
167  14    07  2012          37  37  18  0.2  88.9  12.9  14.6 9  12.5  10.4   

      FWI Classes  
167  fire     NaN

In [13]:
df['Classes'].unique()

array(['not fire', 'fire', 'Classes', nan], dtype=object)

In [14]:
df['Classes'].replace(np.nan,"fire",inplace=True)

In [15]:
df['Classes'].unique()

array(['not fire', 'fire', 'Classes'], dtype=object)

In [16]:
df[df['Classes']=='Classes']

day  month  year  Temperature  RH  Ws  Rain  FFMC  DMC  DC  ISI  BUI  \
123  day  month  year  Temperature  RH  Ws  Rain  FFMC  DMC  DC  ISI  BUI   

     FWI  Classes  
123  FWI  Classes

In [17]:
#Removing the duplicate header

df.drop(index=123, axis=0,inplace=True)

In [18]:
df['Classes'].unique()

array(['not fire', 'fire'], dtype=object)

In [19]:
df[df['day']=='day']

Empty DataFrame
Columns: [day, month, year, Temperature, RH, Ws, Rain, FFMC, DMC, DC, ISI, BUI, FWI, Classes]
Index: []

In [20]:
df[null_mask]

day month  year Temperature  RH  Ws Rain  FFMC   DMC      DC   ISI   BUI  \
167  14    07  2012          37  37  18  0.2  88.9  12.9  14.6 9  12.5  10.4   

      FWI Classes  
167  fire    fire

In [21]:
#To remove the space in b/w the digits like here DC having space in b/w the digits 14.6 9
df['DC']=df['DC'].str.replace(" ","")

In [22]:
df[null_mask]

day month  year Temperature  RH  Ws Rain  FFMC   DMC     DC   ISI   BUI  \
167  14    07  2012          37  37  18  0.2  88.9  12.9  14.69  12.5  10.4   

      FWI Classes  
167  fire    fire

In [23]:
#Replacing the FWI with the average value of FWI having classes as "fire"
df_temp=df.copy()
df_temp.drop(index=167,axis=0,inplace=True)
df_temp[df_temp['FWI']=='fire']

Empty DataFrame
Columns: [day, month, year, Temperature, RH, Ws, Rain, FFMC, DMC, DC, ISI, BUI, FWI, Classes]
Index: []

In [24]:
Avg_FWI_Of_Fire_Classes=df_temp[df_temp['Classes']=='fire']['FWI'].astype(float).median()
df['FWI']=df['FWI'].str.replace('fire',str(Avg_FWI_Of_Fire_Classes))

In [25]:
df[null_mask]

day month  year Temperature  RH  Ws Rain  FFMC   DMC     DC   ISI   BUI  \
167  14    07  2012          37  37  18  0.2  88.9  12.9  14.69  12.5  10.4   

      FWI Classes  
167  10.5    fire

# INFO of Dataset

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 244 entries, 0 to 245
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   day          244 non-null    object
 1   month        244 non-null    object
 2   year         244 non-null    object
 3   Temperature  244 non-null    object
 4   RH           244 non-null    object
 5   Ws           244 non-null    object
 6   Rain         244 non-null    object
 7   FFMC         244 non-null    object
 8   DMC          244 non-null    object
 9   DC           244 non-null    object
 10  ISI          244 non-null    object
 11  BUI          244 non-null    object
 12  FWI          244 non-null    object
 13  Classes      244 non-null    object
dtypes: object(14)
memory usage: 36.7+ KB


In [27]:
df.isnull().sum()

day            0
month          0
year           0
Temperature    0
RH             0
Ws             0
Rain           0
FFMC           0
DMC            0
DC             0
ISI            0
BUI            0
FWI            0
Classes        0
dtype: int64

In [28]:
df.head()

day month  year Temperature  RH  Ws  Rain  FFMC  DMC    DC  ISI  BUI  FWI  \
0  01    06  2012          29  57  18     0  65.7  3.4   7.6  1.3  3.4  0.5   
1  02    06  2012          29  61  13   1.3  64.4  4.1   7.6    1  3.9  0.4   
2  03    06  2012          26  82  22  13.1  47.1  2.5   7.1  0.3  2.7  0.1   
3  04    06  2012          25  89  13   2.5  28.6  1.3   6.9    0  1.7    0   
4  05    06  2012          27  77  16     0  64.8    3  14.2  1.2  3.9  0.5   

    Classes  
0  not fire  
1  not fire  
2  not fire  
3  not fire  
4  not fire

In [29]:
#Converting the Categorical feature into numerical feature with label or onehot encoding

df['Classes']=df['Classes'].map({'fire':1,'not fire':0})
df.head()

day month  year Temperature  RH  Ws  Rain  FFMC  DMC    DC  ISI  BUI  FWI  \
0  01    06  2012          29  57  18     0  65.7  3.4   7.6  1.3  3.4  0.5   
1  02    06  2012          29  61  13   1.3  64.4  4.1   7.6    1  3.9  0.4   
2  03    06  2012          26  82  22  13.1  47.1  2.5   7.1  0.3  2.7  0.1   
3  04    06  2012          25  89  13   2.5  28.6  1.3   6.9    0  1.7    0   
4  05    06  2012          27  77  16     0  64.8    3  14.2  1.2  3.9  0.5   

   Classes  
0        0  
1        0  
2        0  
3        0  
4        0

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 244 entries, 0 to 245
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   day          244 non-null    object
 1   month        244 non-null    object
 2   year         244 non-null    object
 3   Temperature  244 non-null    object
 4   RH           244 non-null    object
 5   Ws           244 non-null    object
 6   Rain         244 non-null    object
 7   FFMC         244 non-null    object
 8   DMC          244 non-null    object
 9   DC           244 non-null    object
 10  ISI          244 non-null    object
 11  BUI          244 non-null    object
 12  FWI          244 non-null    object
 13  Classes      244 non-null    int64 
dtypes: int64(1), object(13)
memory usage: 36.7+ KB


In [31]:
#Changing the data type of each feature based on our requirement

df=df.astype({'day':int, 'month':int, 'year':int, 'Temperature':int, 'RH':int, 'Ws':int, 'Rain':float, 'FFMC':float,
       'DMC':float, 'DC':float, 'ISI':float, 'BUI':float, 'FWI':float, 'Classes':int})

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 244 entries, 0 to 245
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   day          244 non-null    int32  
 1   month        244 non-null    int32  
 2   year         244 non-null    int32  
 3   Temperature  244 non-null    int32  
 4   RH           244 non-null    int32  
 5   Ws           244 non-null    int32  
 6   Rain         244 non-null    float64
 7   FFMC         244 non-null    float64
 8   DMC          244 non-null    float64
 9   DC           244 non-null    float64
 10  ISI          244 non-null    float64
 11  BUI          244 non-null    float64
 12  FWI          244 non-null    float64
 13  Classes      244 non-null    int32  
dtypes: float64(7), int32(7)
memory usage: 30.0 KB


In [32]:
df.head()

day  month  year  Temperature  RH  Ws  Rain  FFMC  DMC    DC  ISI  BUI  \
0    1      6  2012           29  57  18   0.0  65.7  3.4   7.6  1.3  3.4   
1    2      6  2012           29  61  13   1.3  64.4  4.1   7.6  1.0  3.9   
2    3      6  2012           26  82  22  13.1  47.1  2.5   7.1  0.3  2.7   
3    4      6  2012           25  89  13   2.5  28.6  1.3   6.9  0.0  1.7   
4    5      6  2012           27  77  16   0.0  64.8  3.0  14.2  1.2  3.9   

   FWI  Classes  
0  0.5        0  
1  0.4        0  
2  0.1        0  
3  0.0        0  
4  0.5        0

In [33]:
df.to_csv("Cleaned_Data.csv")